Feasibility pump

ZedongPeng opened the pull request:
## Fixes # .

## Summary/Motivation:


## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jsiirola committed:
Update template generation for new constraint rules interface

jsiirola committed:
Merge branch 'master' into constraint-rule-init

jsiirola committed:
Support templates of components with nonfinite index sets.

jsiirola committed:
Merge branch 'master' into constraint-rule-init

jsiirola committed:
Standardize check for empty scalar constraint

jsiirola committed:
Raise error any time an indexed constraint is initialized by a constant

jsiirola committed:
Add the ability to "implicitly" initialize indexed constraints

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

jsiirola committed:
Moving ComponentMap, ComponentSet to pyomo.common.collections

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

jsiirola committed:
Merge branch 'master' into constraint-rule-init

jsiirola committed:
Update all ComponentMap, ComponentSet import locations

jsiirola committed:
Deprecate pyomo.core.kernel.component_{map,set} modules

jsiirola committed:
Remove development import unintentionally added

jsiirola committed:
Updating additional references to component_map and component_set

emma58 committed:
Adding option to FME to perform only integer operations

emma58 committed:
Oops, added test doesn't need a solver

mrmundt committed:
Push rather than PR

jsiirola committed:
Rebuild GHA download caches (bump cache version)

jsiirola committed:
Merge pull request #1598 from jsiirola/bump-cache-ver

Rebuild GHA download caches (bump cache version)

jsiirola committed:
Merge branch 'master' into constraint-rule-init

jsiirola committed:
Merge branch 'master' into move-componentmap

emma58 committed:
Fixing a bug in FME where we added to the list of variables to eliminate more than once for the same variable. This changed the tests because as a result we were reordering constraints unncessarily after doing all the projections

jsiirola committed:
Removed unreferenced module object; fixing typo

jsiirola committed:
Merge branch 'master' into integer-arithmetic-fme

jsiirola committed:
Merge pull request #1593 from jsiirola/move-componentmap

Move ComponentMap, ComponentSet to pyomo.common.collections

emma58 committed:
Adding an option for how small a float to consider 0 while performing FME, also adding verbose logging to FME

emma58 committed:
Merge branch 'integer-arithmetic-fme' of https://github.com/emma58/pyomo into integer-arithmetic-fme

emma58 committed:
Merge branch 'master' into integer-arithmetic-fme

emma58 committed:
Discovering the to_expression method for standard repn, which changes baselines again. Also enforcing integer arithmetic throughout, but not quite successfully yet

emma58 committed:
Fixing bug where the constant wasn't multiplied in scalar multiplication

emma58 committed:
Cleaning up since I put tolerances on the class, adding one test for non-integer errors

jsiirola committed:
Resolve quotes when parsing /etc/os-release

jsiirola committed:
Fixing test failure message

blnicho committed:
Merge pull request #1592 from jsiirola/constraint-rule-init

Rework constraint initialization

blnicho committed:
Merge pull request #1603 from jsiirola/os-release-parse-fix

Resolve quotes when parsing /etc/os-release

ZedongPeng committed:
fix no bound provided when solving rNLP using IPOPT

Robbybp committed:
propagate options to deactivate_model_at

Robbybp committed:
test for error with allow_skip=False

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

jsiirola committed:
Add support for class method Initializers

jsiirola committed:
PY3 portability fix

jsiirola committed:
(re)Adding IndexedSet.data() method

michaelbynum committed:
pynumero matvec performance improvements

ZedongPeng committed:
add Best solution found time and fix rNLP bound update

jsiirola committed:
Reverting a test to its pre-Set-rewrite state

michaelbynum committed:
pynumero mpi matvec performance

michaelbynum committed:
pynumero mpi matvec performance

michaelbynum committed:
pynumero mpi matvec performance

carldlaird committed:
Merge pull request #1608 from jsiirola/instancemethod-initializer

Add support for class method Initializers

michaelbynum committed:
pynumero matvec performance and bug

michaelbynum committed:
pynumero mpi block vector performance

michaelbynum committed:
pynumero mpi block vector performance

michaelbynum committed:
pynumero bug

blnicho committed:
Merge pull request #1609 from jsiirola/indexedset-data

(re)Adding IndexedSet.data() method

michaelbynum committed:
handling empty block rows in mpi matvec

ZedongPeng committed:
add get_main_elapsed_time import in single_tree

michaelbynum committed:
mpi matvec - handling empty rows

ZedongPeng committed:
feasibility pump useful function

ZedongPeng committed:
Merge branch 'new-lp-nlp' into feasibility_pump

ZedongPeng committed:
add new functions for feas_pump

ZedongPeng committed:
remove redundant code

ZedongPeng committed:
remove redundant code

ZedongPeng committed:
add necessary imports

emma58 committed:
Clearing the global name buffer at the end

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,330 @@
+    return distance <= config.integer_tolerance
+
+
+def solve_feas_pump_NLP_subproblem(solve_data, config):
```
This function is practically the same as the `solve_NLP_subproblem` function, excepting the part Defining the objective. It will get harder to maintain this code. Couldn't we just use the `solve_NLP_subproblem` with a flag and that's it? Besides, I do not remember if in the paper we enforce the objective cut to the NLP too. Do we?

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,330 @@
+                             config)
+        fix_nlp, fix_nlp_results = solve_NLP_subproblem(
+            solve_data, config)
+        assert fix_nlp_results.solver.termination_condition is tc.optimal, 'Feasibility pump fix-nlp subproblem not optimal'
```
fixed_nlp instead of fix-nlp

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,330 @@
+    if solve_data.solution_improved:
+        solve_data.best_solution_found = solve_data.working_model.clone()
+        assert is_feasible(solve_data.best_solution_found, config), \
+            "Best found model infeasible! There might be a problem with the precisions - the feaspump seems to have converged (error**2 <= integer_tolerance). " \
```
Best found "solution" I guess. And I don't understand when can this ever happen.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,330 @@
+
+        # Solve NLP subproblem
+        # The constraint linearization happens in the handlers
+        fixed_nlp, fixed_nlp_result = solve_feas_pump_NLP_subproblem(
```
inside of this function you have a way to handle the infeasible NLP subproblem with an assert, but then you call the functions below again. Is this right?

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -33,7 +33,7 @@ def _get_GDPopt_config():
     CONFIG.declare("strategy", ConfigValue(
         default="OA",
-        domain=In(["OA", "GBD", "ECP", "PSC", "GOA"]),
+        domain=In(["OA", "GBD", "ECP", "PSC", "GOA", "feas_pump"]),
```
This is something that we need to thing more. We want to have the feasibility pump as a standalone strategy and also as an initialization procedure. Contrary to the initialization strategies we have in place (rNLP, max_binary) that decide how to deal with the first iteration, we also want to decide if we want to run the feasibility pump before starting OA or ECP. At the very least we can use the initialization, but the paper I wrote shows that the cuts generated in the process are also useful (we can decide this with an option).

bernalde reviewed (commented):
This is looking good so far! My main comment has to do with the way that we plan to use the feasbility pump. It should be both a standalone strategy (as OA or ECP) but also an initialization procedure to provide a) a feasible starting point to the other strategies, or b) provide both the initial point and the generated cuts (special attention to the `fp_transfer_cuts` option in GAMS/DICOPT. There are also some options which are still missing for the feasibility pump. Use the references of our implementation in DICOPT to guide yourself https://www.gams.com/latest/docs/S_DICOPT.html#DICOPT_fp

Robbybp committed:
Explain argumnets in docstring

michaelbynum committed:
pynumero performance

ZhifeiYL commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -66,38 +66,33 @@ def solve_OA_master(solve_data, config):
-        MindtPy.MindtPy_oa_obj = Objective(
-            expr=main_objective.expr + MindtPy.MindtPy_penalty_expr,
-            sense=main_objective.sense)
+    if config.strategy is 'feas_pump':
```
I think it's probably better to use a flag for 'feas_pump' with a default value set to be False if we want a standalone version of feas_pump.

emma58 committed:
More sensible mapping to store bigM value and source. Use None as source if calculated and make second argument of tuple (M_lower, M_upper)

jsiirola committed:
Add deprecated support for Var(within=RealSet)

samryan18 committed:
Remove truthiness assumption in TuplizeValuesInitializer

TuplizeValuesInitializer should not assume that the sequence can be truthy (e.g. pandas indices do not support this)

See: https://pandas.pydata.org/pandas-docs/version/0.15/gotchas.html

ZedongPeng committed:
add feasibility_pump options

ZedongPeng committed:
change solve_OA_master to solve_MIP_master

ZedongPeng committed:
set feas_pump strategy options

samryan18 committed:
Import pandas and create pandas_available flag

samryan18 committed:
Add test pandas multi-index set init

samryan18 committed:
Fix line length

samryan18 committed:
Use pyomo common dependencies for pandas_available

samryan18 committed:
Update test_set.py

samryan18 committed:
test that dimension is inferred correctly

samryan18 committed:
Remove loop for clarity

kaklise committed:
graphics updates for seaborn 0.11

kaklise committed:
Kept distplot for older verisons of seaborn

kaklise committed:
More updates for py35

kaklise committed:
minor update

samryan18 committed:
Fix hyperlink in contribution guide

samryan18 committed:
Tiny typo

kaklise committed:
minor update

jsiirola committed:
Make _check_version a public method compatible with deferred imports

jsiirola committed:
Switch between distplot/histplot based on seaborn version

jsiirola committed:
Fixing typo in module name

jsiirola committed:
Fix standard repn of external functions with fixed arguments

ZhifeiYL reviewed (commented):
Reviewed.

ZedongPeng committed:
backup

emma58 committed:
BigM source dictionary stores separate sources for both lower and upper values, adding method to just get values, updating tests

emma58 committed:
Removing unused import

emma58 committed:
Deprecating the old get_m_value_src method since I renamed it anyway...

jsiirola committed:
Merge pull request #1622 from kaklise/parmest-graphics

Updated parmest graphics due to recent changes in seaborn

jsiirola committed:
Merge branch 'master' into fix-standard-repn-external-fcn

jsiirola committed:
Merge branch 'master' into fix-realset

jsiirola committed:
Merge branch 'master' into bigm-src-dict-fix

jsiirola committed:
Merge branch 'master' into matvec

jsiirola committed:
Merge branch 'master' into dae

jsiirola committed:
Merge branch 'master' into integer-arithmetic-fme

jsiirola committed:
Merge branch 'master' into patch-1

ZedongPeng committed:
change iteration_limit to NonNegativeInt

ZedongPeng committed:
backup

ZedongPeng committed:
backup

ZedongPeng committed:
remove redundant code

ZedongPeng committed:
backup

ZedongPeng committed:
first working version of feas_pump

ZedongPeng committed:
fix bug and organize configs

bernalde commented on pyomo/contrib/mindtpy/MindtPy.py:
```diff
@@ -79,6 +79,12 @@ def solve(self, model, **kwds):
+        solve_data.timing = Container()
+        solve_data.curr_int_sol = []
+        solve_data.prev_int_sol = []
+
```
I really like this!

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -5,15 +5,10 @@
 
 def _get_GDPopt_config():
```
Do you think we should change the name of this util function `to _get_GDPopt_MindtPy_config()` ?

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -237,10 +155,142 @@ def _get_GDPopt_config():
+        default=1E-2,
+        domain=PositiveFloat,
+        description="Optimality tolerance (relative gap) to use for solving MIP projection problem"
+    ))
```
There is a missing option here (which is in DICOPT too but is undocumented named `fp_softcuts`;    whether to add linearization cuts as soft constraints (penalized slack variable) to MIP relaxation. This is a heuristic for non-convex and we leave it off by default.
After reading the whole code, I think that this should be controlled by the single option `add_slacks` but make those slacks available to feaspump too.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+"""Generates objective functions like L1, L2 and Linf distance"""
```
This comment should go elsewhere.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+from pyomo.contrib.mindtpy.mip_solve import solve_MIP_master, handle_master_mip_optimal
+
+
+def feas_pump_converged(solve_data, config):
```
I think there is a confusion here. In feaspump you only compute the distance with respect to the integer variables, not all of them. Although I like having the flexibility for both. Can you pass a flag to this function which by default computed the difference only of the discrete variables?

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+        sub_nlp.component_data_objects(Objective, active=True))
+    main_objective.deactivate()
+
+    # TODO: need to comfirm with David, whether to add increasing_objective_cut for FP-NLP
```
Great point! Yes, since the original objective is not considered in the objective here, we add the bounding constraint, but to avoid making this problem infeasible, we do not include the delta.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+    MindtPy.feas_pump_nlp_obj = generate_L2_objective_function(
+        sub_nlp,
+        solve_data.mip,
+        discretes_only=True
```
`discrete_only` and perfect!

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+    # if OA-like or feas_pump converged, update Upper bound,
+    # add no_good cuts and increasing objective cuts (feas_pump)
+    if feas_pump_converged(solve_data, config):
+        # TODO: Need to think about the efficiency of warm start, use solve_data.mip or sub_nlp
```
It is always good to initialize the NLP models using the MILP solution. Even in the feaspump.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+        solve_data.mip_subiter = 0
+        # solve MILP master problem
+        feas_mip, feas_mip_results = solve_MIP_master(
+            solve_data, config, feas_pump=True)
```
We should try to unify the way we refer to the feasibility pump. Either `feaspump`, `feas_pump`, or `fp`, but only one.

ZedongPeng replied:
For the configurations, I want to keep the name of options in MindtPy the same as Dicopt. So I use `fp`. However, for the functions and flags of feasibility pump in MindtPy, I think `fp `is too concise. So I use `feas_pump`. And for the output of loggers, I use the full name of `feasibility pump`.

bernalde replied:
Alright, but now you added a `feas_mip`. Let's work out a good naming here.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,186 @@
+        elif feas_mip_results.solver.termination_condition is tc.infeasible:
+            config.logger.info('FP-MIP infeasible')
+            break
+        # elif feas_mip_results.solver.termination_condition is tc.maxIterations:
```
If this is the case, we can just use the incumbent solution of the MIP (if any) and keep going after raising a warning.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -237,10 +155,142 @@ def _get_GDPopt_config():
+        description="Tolerance on when to consider optimal value of projection problem as zero, which may trigger the solution of a Sub-NLP"
+    ))
+    CONFIG.declare("fp_mipgap", ConfigValue(
+        default=1E-2,
```
We need to be careful here, if using 'feaspump' only then reduce this gap to zero.

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -66,38 +65,35 @@ def solve_OA_master(solve_data, config):
+            discretes_only=True)
+    else:
+        if config.add_slack:
+            MindtPy.del_component('MindtPy_penalty_expr')
```
We can add slacks to feaspump too in the soft_cuts version of it.

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -156,8 +158,20 @@ def handle_NLP_subproblem_optimal(fixed_nlp, solve_data, config):
-    if config.strategy == 'OA':
+        # add obj increasing constraint for feas_pump
+        if feas_pump:
+            if solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.find_component('increasing_objective_cut') is not None:
```
Rename `increasing_objective_cut` as `improving_objective_cut`

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -156,8 +158,20 @@ def handle_NLP_subproblem_optimal(fixed_nlp, solve_data, config):
+                    'increasing_objective_cut')
+            if main_objective.sense == minimize:
+                solve_data.mip.MindtPy_utils.MindtPy_linear_cuts.increasing_objective_cut = Constraint(expr=solve_data.mip.MindtPy_utils.objective_value
+                                                                                                       <= solve_data.UB - config.feas_pump_delta*min(1e-4, abs(solve_data.UB)))
```
This 1e-4 should be an extra parameter called like `fp_min_improvement`

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_feas_pump.py:
```diff
@@ -0,0 +1,157 @@
+"""Tests for the MindtPy solver plugin."""
```
We need to activate these tests. Besides @Justinyl has some examples by hand that you could use.

bernalde replied:
Did you ever receive the example files?

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,60 @@ def var_bound_add(solve_data, config):
+                                            zip(model.component_data_objects(Var),
+                                                setpoint_model.component_data_objects(Var))))
+    assert len(model_vars) == len(
+        setpoint_vars), "Trying to generate L1 objective function for models with different number of variables"
```
Norm2 in this case

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,60 @@ def var_bound_add(solve_data, config):
+    return Objective(expr=(
+        sum([(model_var - setpoint_var.value)**2
+             for (model_var, setpoint_var) in
+             zip(model_vars, setpoint_vars)])))
```
Here you are never computing the squared root. I would change the function documentation and even the function name to `generate_Norm2sq_objective_function`

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,60 @@ def var_bound_add(solve_data, config):
                         var.setub(config.continuous_var_bound)
+
+
+def generate_L2_objective_function(model, setpoint_model, discretes_only=False):
```
For this whole file, I think we can be a bit more explicit with the function names. Let's use Norm1 and Norm2 instead of L (it is not clear what we mean with L).

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,60 @@ def var_bound_add(solve_data, config):
+        obj_blk.L1_obj_ub_constr[c_ub].set_value(
+            expr=v_model - v_setpoint.value <= obj_blk.L1_obj_var)
+
+    return Objective(expr=obj_blk.L1_obj_var)
```
There is a NormInfty function missing, right?

ZedongPeng replied:
I didn't add the NormInfty function here since I didn't see Norm inf is used in the paper. I can add this if we need it.

bernalde replied:
We are thinking about the upcoming paper too!

bernalde commented on pyomo/contrib/mindtpy/nlp_solve.py:
```diff
@@ -156,8 +158,20 @@ def handle_NLP_subproblem_optimal(fixed_nlp, solve_data, config):
+                                                                                                       >= solve_data.LB + config.feas_pump_delta*min(1e-4, abs(solve_data.LB)))
+
+            # Add the linear cut
+    if config.strategy == 'OA' or feas_pump:
```
Just a general question. We could potentially add affine cuts to the feaspump (I don't know if that is intelligent). If that is the case how should the code look like?
In line 189 below the comment needs to be updated to `no_good_cuts`.

ZedongPeng replied:
Add `add_affine_cuts` option

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -123,48 +83,16 @@ def _get_GDPopt_config():
-        default=1E-8,
-        description="Tolerance on variable equal to zero."
-    ))
     CONFIG.declare("initial_feas", ConfigValue(
```
Is this the feasibility pump? If so, leave it `False` by default and change the name to a more complete `initial_feaspump`.

ZedongPeng replied:
I checked the code. This is an ancient option created before the feasibility pump branch. It's somehow useless, so I deleted it.

bernalde reviewed (commented):
Great progress, I made a lot of comments along the way. Most are cosmetic and shouldn't take a lot of effort. As examples, we could either use the example that @Justinyl will send you or use the examples in the Bonami paper https://link.springer.com/article/10.1007/s10107-008-0212-2
This is looking better every time I look at it.

pareyesv committed:
doc: fix variable name error

ZedongPeng committed:
solve David's comment

ZedongPeng committed:
fix bug

samryan18 committed:
Update contribution_guide.rst

samryan18 committed:
Match surrounding line length

Not sure what the benefit is from limiting rst line-length is but 🤷

samryan18 committed:
Minor documentation changes

Really just kicking off GH workflows again to see if pip/SSL issues were temporary

ZedongPeng committed:
change L1 to Norm1

blnicho committed:
Fixing a few typos

blnicho committed:
Merge pull request #1623 from jsiirola/fix-standard-repn-external-fcn

Fix standard repn of external functions with fixed arguments

blnicho committed:
Merge pull request #1620 from samryan18/patch-1

Remove truthiness assumption for sequence in TuplizeValuesInitializer

blnicho committed:
Merge pull request #1606 from Robbybp/dae

Error and warning options in solve_consistent_initial_conditions

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -57,16 +55,16 @@ def solve_feas_pump_NLP_subproblem(solve_data, config):
     main_objective.deactivate()
 
-    # TODO: need to comfirm with David, whether to add increasing_objective_cut for FP-NLP
+    # TODO: need to comfirm with David, whether to add improving_objective_cut for FP-NLP
```
Yes we add this cut for the NLP, but without the delta.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -195,17 +195,17 @@ def generate_L2_objective_function(model, setpoint_model, discretes_only=False):
                                                 setpoint_model.component_data_objects(Var))))
     assert len(model_vars) == len(
-        setpoint_vars), "Trying to generate L1 objective function for models with different number of variables"
+        setpoint_vars), "Trying to generate Norm2 objective function for models with different number of variables"
```
Squared Norm2

emma58 committed:
Renaming get_M_from_suffixes, adding to my deprecation error message so as not to annoy myself.

emma58 committed:
Merge branch 'bigm-src-dict-fix' of https://github.com/emma58/pyomo into bigm-src-dict-fix

ZedongPeng committed:
comments fixed

ZedongPeng committed:
update add_nogood_cuts doc

emma58 committed:
Restoring logging level after transformation, using user-specified logging level to enable verbose output also.

emma58 committed:
Merge branch 'integer-arithmetic-fme' of https://github.com/emma58/pyomo into integer-arithmetic-fme

emma58 committed:
Merge branch 'master' into integer-arithmetic-fme

jsiirola committed:
Disable the OS package cache for OSX

blnicho committed:
Fixing typo

blnicho committed:
Merge pull request #1619 from jsiirola/fix-realset

Add deprecated support for Var(within=RealSet)

blnicho committed:
Merge pull request #1624 from pareyesv/patch-1

doc: fix variable name error

samryan18 committed:
Add note on good commits

emma58 committed:
Adding name_buffer to add_slacks transformation, switching to getname

michaelbynum committed:
better handling of operations on MPIBlockVector when running in serial

ZedongPeng committed:
add Norm infinity and fix the bug in Norm1

ZedongPeng committed:
fix spelling error

ZedongPeng committed:
feasibility_pump1 example work

ZedongPeng committed:
update log outputs

ZedongPeng committed:
debugging

jsiirola committed:
Add additional '_available' flags to doctest env setup

jsiirola committed:
Skip units doctests if pint is not available

jsiirola committed:
Merge pull request #1618 from emma58/bigm-src-dict-fix

GDP: consistent key for calculated M values in bigM dictionary

jsiirola committed:
Fixing indentation in doctest :skipif:

jsiirola committed:
Merge pull request #1625 from jsiirola/disable-osx-os-cache

Disable the OS package cache for OSX

blnicho committed:
Merge pull request #1621 from samryan18/patch-2

Update contribution guide

blnicho committed:
Merge branch 'master' into matvec

jsiirola committed:
FME: correctly round coefs when within integer_tolerance

Also, avoid generating error message strings until we know that we need
to emit them.

blnicho committed:
Fixing typo

jsiirola committed:
NFC: Removing trailing whitespace

jsiirola committed:
Merge pull request #1594 from emma58/integer-arithmetic-fme

Add integer arithmetic option to FME transformation and add tolerance for floating point version

blnicho committed:
Updating the CHANGELOG

jsiirola committed:
Merge branch 'master' into finalize-release

jsiirola committed:
Setting deprecation warning version information

jsiirola committed:
Updating the changelog

jsiirola committed:
Finalizing the 5.7.1 release

blnicho committed:
Merge pull request #1628 from Pyomo/finalize-release

Finalizing the 5.7.1 release

blnicho committed:
Fixing deprecation tests

michaelbynum committed:
Merge pull request #1610 from michaelbynum/matvec

Pynumero MPI Matvec Performance

jsiirola committed:
Merge pull request #1629 from blnicho/fix-deprecation-tests

Fixing deprecation decorator tests

blnicho committed:
Updating CHANGELOG

blnicho committed:
Updating the CHANGELOG

blnicho committed:
Merge pull request #1630 from Pyomo/finalize-release

Adding a couple more entries to the CHANGELOG for the 5.7.1 release

jsiirola committed:
Preparing master for development (5.7.2.dev0)

jsiirola committed:
Merge pull request #1631 from Pyomo/finalize-release

Preparing master for development (5.7.2.dev0)

jsiirola committed:
Add a cmake option to disable building all HSL interfaces

jsiirola committed:
Merge pull request #1632 from jsiirola/pynumero-disable-hsl

pynumero: add a cmake option to disable building all HSL interfaces

ZedongPeng committed:
remove assert is_feasible

ZedongPeng committed:
change sub_nlp to fp_nlp, add improving_objective_cut to fp_nlp

ZedongPeng committed:
fix several bugs for FP-OA

emma58 committed:
Merge branch 'master' into add-slacks-performance

emma58 committed:
Only get constraint name once

jsiirola committed:
Merge pull request #1637 from emma58/add-slacks-performance

Use name buffer in add slacks transformation

ZedongPeng committed:
copy value bug fixed

DLWoodruff committed:
fix two bugs in parmest and add (edited) example

DLWoodruff committed:
add a test of return values

DLWoodruff committed:
added a little more documentation

DLWoodruff committed:
cleanup the top of retvals.py

jsiirola committed:
Lengthing the ctypes.CDLL load timeout to 10 seconds

DLWoodruff committed:
dropping the doctest

blnicho committed:
Fixing typo

blnicho committed:
Merge pull request #1644 from jsiirola/cdll-timeout

Increase CtypesEnviron ctypes.CDLL load timeout to 10 seconds

mrmundt committed:
Merge branch 'master' of https://github.com/Pyomo/pyomo

mrmundt committed:
Adding python-louvain to testing suite, as per issue 1696

DLWoodruff committed:
return_values example with doctest

DLWoodruff committed:
add the blank line after doctest line

DLWoodruff committed:
making the doctest condional on having ipopt

blnicho committed:
Fixing a whitespace issue with the doctests

blnicho committed:
Skip doctest when pandas is not available

DLWoodruff committed:
remove the retvals example because it is entirely in the docs

DLWoodruff committed:
Merge branch 'retvalbug' of https://github.com/DLWoodruff/pyomo into retvalbug

ZedongPeng committed:
change is_binary to is_integer, add FP-OA test

blnicho committed:
Merge pull request #1645 from DLWoodruff/retvalbug

parmest return_values bug fix

blnicho committed:
Merge pull request #1646 from mrmundt/issue-1626

Add python-louvain to testing infrastructure

supermihi committed:
Fix a typo in gurobi_direct.py

blnicho committed:
Merge pull request #1649 from supermihi/patch-1

Fix a typo in gurobi_direct.py

blnicho committed:
Adding public API for checking the `mutable` flag on a Param component

blnicho committed:
Merge branch 'master' into param-mutability-check

ZedongPeng committed:
add no_cycling_cuts for feasibility pump

ZedongPeng committed:
add fp_transfercuts

ZedongPeng committed:
change no_cycling_cuts to fp_no_cycling_cuts

ZedongPeng committed:
change integer_cuts to nogood_cuts

ZedongPeng committed:
deactivate the last nogood cut

blnicho committed:
Merge pull request #1651 from blnicho/param-mutability-check

Add public API for checking Param mutability

mrmundt committed:
Changing to environment files re: security risk

mrmundt committed:
Env is not acting as expected

mrmundt committed:
Moving piping

mrmundt committed:
Missing one thing

mrmundt committed:
Character issue

mrmundt committed:
Changing LIBRARY_PATH invocations

mrmundt committed:
Changing LIBRARY_PATH invocations

mrmundt committed:
Failures in MPI but nowhere else

mrmundt committed:
Adding changes to pr masterg

mrmundt committed:
Name of GHA setup-miniconda slightly changed

jsiirola committed:
Merge pull request #1654 from mrmundt/env_files

Address GitHub Actions Security Risk in Workflow

jsiirola committed:
cplex_direct: explicitly open (and close) log files for cplex>=12.5.1

jsiirola committed:
Merge branch 'master' into cplex-log-fix

jsiirola committed:
Adding missing six import

jsiirola committed:
Fixing typo

michaelbynum committed:
Merge pull request #1655 from jsiirola/cplex-log-fix

Explicitly close log files for cplex_direct

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -124,18 +131,22 @@ def init_rNLP(solve_data, config):
                                  solve_data.mip.MindtPy_utils.variable_list,
                                  config, ignore_integrality=True)
+            if config.init_strategy == 'feas_pump':
+                # TODO：remove here
```
This line is making the tests in OSX fail with error `SyntaxError: Non-ASCII character '\xef' in file /Users/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/initialization.py on line 146, but no encoding declared; see http://python.org/dev/peps/pep-0263/ for details`
There is a weird symbol in there that makes our life hard

bernalde commented on pyomo/contrib/mindtpy/MindtPy.py:
```diff
@@ -98,19 +104,19 @@ def solve(self, model, **kwds):
+            config.iteration_limit = 0
+            config.add_nogood_cuts = True
+        if config.init_strategy == "feas_pump":
+            solve_data.fp_iter = 1
```
I don't know if this is right. Even if we use the feaspump as an initialization strategy, the iterations might be more than 1. Actually, in DICOPT I think we set the default to 20.

ZedongPeng replied:
Here `solve_data.fp_iter` is a counter.

bernalde replied:
Why do we initialize it then?

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -64,41 +59,6 @@ def _get_GDPopt_config():
-        doc="Which MIP subsolver options to be passed to the solver while "
-            "solving the mixed-integer master problems."
-    ))
     CONFIG.declare("call_after_master_solve", ConfigValue(
```
This and `call_after_nlp_solve` should have their own timers too.

bernalde commented on pyomo/contrib/mindtpy/config_options.py:
```diff
@@ -237,10 +150,141 @@ def _get_GDPopt_config():
+        description="Additional relative decrement of cutoff value for the original objective function"
+    ))
+    CONFIG.declare("fp_iteration_limit", ConfigValue(
+        default=20,
```
See other comment on setting 1 iteration limit when initializing with feaspump.

bernalde commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -332,6 +332,8 @@ def add_ecp_cuts(target_model, solve_data, config,
 def add_nogood_cuts(var_values, solve_data, config, feasible=False):
     """
     Adds integer cuts; modifies the model to include integer cuts
+    This adds an integer cut to the feasible_nogood_cuts ConstraintList, which is not activated by default.
```
Replace "integer cuts" by "no good cuts" in the description.

bernalde commented on pyomo/contrib/mindtpy/cut_generation.py:
```diff
@@ -375,14 +377,14 @@ def add_nogood_cuts(var_values, solve_data, config, feasible=False):
+    #     MindtPy.MindtPy_linear_cuts.nogood_cuts.add(expr=int_cut)
     # else:
-    #     MindtPy.MindtPy_linear_cuts.feasible_integer_cuts.add(expr=int_cut)
+    #     MindtPy.MindtPy_linear_cuts.feasible_nogood_cuts.add(expr=int_cut)
```
Do we really need these comments here? We can remove them.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,213 @@
+                        solve_data.mip.MindtPy_utils.variable_list)
+                    if (not discrete_only) or milp_var.is_integer()))
+
+    return distance <= config.integer_tolerance
```
I think that this convergence is a little harder than it should be. Since you are comparing the integer tolerance of a single variable against the quadratic sum of all discrete variables. Should we compare it against the maximum difference? Or the average? What do you think?

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,213 @@
+            solve_data, config)
+        main_objective = next(
+            fixed_nlp.component_data_objects(Objective, active=True))
+        # assert fixed_nlp_results.solver.termination_condition is tc.optimal, 'Feasibility pump fixed_nlp subproblem not optimal'
```
This comment can be removed.

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,213 @@
+            config.logger.warning('FP-MIP reaches max TimeLimit')
+        elif feas_mip_results.solver.termination_condition is tc.infeasible:
+            config.logger.warning('FP-MIP infeasible')
+            # TODO: needs to be checked here.
```
What needs to be checked?

bernalde commented on pyomo/contrib/mindtpy/feasibility_pump.py:
```diff
@@ -0,0 +1,213 @@
+
+def add_cycling_cuts(solve_data, config):
+    """
+    add cuts to avoid cycling when the independence constraint qualification is not satisfied.
```
Please add the documentation of this function as the other functions we have.

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -42,7 +44,10 @@ def MindtPy_initialize_master(solve_data, config):
-    if config.strategy == 'OA':
+    if config.init_strategy == 'feas_pump':
+        MindtPy.MindtPy_linear_cuts.fp_no_cycling_cuts = ConstraintList(
+            doc='Cycling cuts')
```
There is a mismatch in the naming between `cycling` and `no_cycling` here. Let's settle with `orthogonality` cuts.

bernalde commented on pyomo/contrib/mindtpy/initialization.py:
```diff
@@ -124,18 +131,22 @@ def init_rNLP(solve_data, config):
             m.dual[c] for c in MindtPy.constraint_list) if config.use_dual else None
         # Add OA cut
-        if main_objective.sense == minimize:
+        if main_objective.sense == minimize and not math.isnan(results['Problem'][0]['Lower bound']):
```
I would add a comment here mentioning that this covers the case when the Lower bound does not exist.

bernalde commented on pyomo/contrib/mindtpy/iterate.py:
```diff
@@ -3,7 +3,7 @@
 from pyomo.contrib.mindtpy.cut_generation import (add_oa_cuts, add_ecp_cuts)
 
-from pyomo.contrib.mindtpy.mip_solve import (solve_OA_master,
+from pyomo.contrib.mindtpy.mip_solve import (solve_MIP_master,
```
we have `solve_MIP_master` and `handle_mip`. Let's settle for the notation. I'm open to suggestions, but maybe let's consider naming them just master and subproblem (and drop the MIP and NLP since it won't be true in general).

bernalde commented on pyomo/contrib/mindtpy/mip_solve.py:
```diff
@@ -17,18 +17,17 @@
 from pyomo.repn import generate_standard_repn
-
 from pyomo.common.dependencies import attempt_import
+from pyomo.contrib.mindtpy.util import generate_Norm1_objective_function
```
Let's avoid function names with a combination of camelCase and this_naming_style.

bernalde commented on pyomo/contrib/mindtpy/single_tree.py:
```diff
@@ -289,7 +289,7 @@ def add_lazy_nogood_cuts(self, var_values, solve_data, config, opt, feasible=Fal
         #                if value(abs(v)) <= int_tol) >= 1)
 
-        # MindtPy.MindtPy_linear_cuts.integer_cuts.add(expr=int_cut)
+        # MindtPy.MindtPy_linear_cuts.nogood_cuts.add(expr=int_cut)
```
I think we can remove these commented lines.

bernalde commented on pyomo/contrib/mindtpy/tests/feasibility_pump2.py:
```diff
@@ -0,0 +1,37 @@
+"""Example 1 in paper "A Feasibility Pump for mixed integer nonlinear programs"
```
Example 2 I guess because the other file had example 1

bernalde commented on pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py:
```diff
@@ -157,7 +157,7 @@ def test_OA_OnlineDocExample(self):
     #                   mip_solver=required_solvers[1],
     #                   nlp_solver=required_solvers[0],
-    #                   add_integer_cuts=True,
+    #                   add_nogood_cuts=True,
```
Why is this test commented out?

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,84 @@ def var_bound_add(solve_data, config):
                         var.setub(config.continuous_var_bound)
+
+
+def generate_Norm2sq_objective_function(model, setpoint_model, discrete_only=False):
```
let's use `norm` instead of `Norm`

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,84 @@ def var_bound_add(solve_data, config):
+
+def generate_Norm2sq_objective_function(model, setpoint_model, discrete_only=False):
+    """Generate objective (FP-NLP subproblem) for minimum euclidean distance to setpoint_model
+    L2 distance of (x,y) = \sqrt{\sum_i (x_i - y_i)^2}
```
This documentation needs to be completed and corrected. Remember we are dealing with the squared two-norm.

bernalde commented on pyomo/contrib/mindtpy/util.py:
```diff
@@ -181,3 +181,84 @@ def var_bound_add(solve_data, config):
+    """Generate objective (FP-NLP subproblem) for minimum euclidean distance to setpoint_model
+    L2 distance of (x,y) = \sqrt{\sum_i (x_i - y_i)^2}
+    discrete_only -- only optimize on distance between the discrete variables
+    TODO: remove setpoint_model
```
What are the `setpoint_model`?

bernalde reviewed (commented):
Great work. I left many comments that are easy to address and correct. After making those changes (and hopefully, we get the other PR merged) let's merge Pyomo master and see how we do on tests.
I think we are still missing some discussion on the feaspump as an initialization method and please read the storm of feasibility pump paper such that we can think of further improvements that can be made.

ZedongPeng committed:
add timer for call_after_master_solve and call_after_subproblem_solve

ZedongPeng committed:
replace integer cuts with no good cuts

ZedongPeng committed:
remove redundant comments

ZedongPeng committed:
calculate maximum difference when checking FP convergence

ZedongPeng committed:
update documentation

ZedongPeng committed:
rename cycling cuts to orthogonality cuts

ZedongPeng committed:
add comments

ZedongPeng committed:
rename solve_MIP_master to solve_master

ZedongPeng committed:
rename handle_master_mip_optimal to handle_master_optimal

ZedongPeng committed:
rename handle_master_mip_other_conditions to handle_master_other_conditions

ZedongPeng committed:
rename handle_master_mip_infeasible to handle_master_infeasible

ZedongPeng committed:
renaming

ZedongPeng committed:
renaming

ZedongPeng committed:
renaming

ZedongPeng committed:
remove redundant comments

ZedongPeng committed:
add documentation for objective generation function

ZedongPeng committed:
add square for integer tolerance

ZedongPeng committed:
fix tolerance

ZedongPeng committed:
keep fp_orthogonality_cuts

ZedongPeng committed:
add add_affine_cuts option

ZedongPeng committed:
rename feas_mip to feas_master

ZedongPeng committed:
rename feas_nlp to feas_subproblem

ZedongPeng committed:
Merge branch 'master' into feasibility_pump

ZedongPeng committed:
add fp_discrete_only option

ZedongPeng committed:
integrate fp_projcuts option

ZedongPeng committed:
add fp_master_norm option

ZedongPeng committed:
add norm constraint to FP-NLP

ZedongPeng committed:
add utf-8

bernalde commented:
You might want to add the utf-8 to the file `/Users/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/tests/feasibility_pump1.py` since it is making the tests in OSX fail. In fact, you might want to add `# -*- coding: utf-8 -*-` to all the files we have.

ZedongPeng committed:
add coding: utf-8 for all files
